HTTP_API.md

```diff
@@ -0,0 +1,952 @@
+
+
+### Core
+
+* #### `AuthPostHandler`
+
+    ###### synopsis
+    Attempts user authenticaiton;
+    used as a simple way to check a user authentication header.
+
+    ###### request
+    `POST /api/v1/auth`
+
+    The request should include an `Authorization` header to be checked.
+
+    ###### response
+    `200` - No body. Getting this response means that the header is valid.
+
+    `401` - User authentication failed. The body is an error message.
+    Getting this response means that the header is invalid.
+
+    `500` - There was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl -X POST --header "Authorization: Basic dXNlcjpwYXNzCg==" localhost:8181/api/v1/auth
+    ```
+
+
+* #### `ClientUrlGetHandler`
+
+    ###### synopsis
+    Returns a URL that a client can connect to, to access Container JFR's
+    WebSocket command channel (see [COMMANDS.md](COMMANDS.md)).
+
+    ###### request
+    `GET /api/v1/clienturl`
+
+    ###### response
+    `200` - The body is `{"clientUrl":"$URL"}`.
+
+    `500` - The URL could not be constructed due to an error with the socket
+    or the host. Or there was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/api/v1/clienturl
+    {"clientUrl":"ws://0.0.0.0:8181/api/v1/command"}
+    ```
+
+
+* #### `GrafanaDashboardUrlGetHandler`
+
+    ###### synopsis
+    Returns the URL of the Grafana dashboard that Container JFR
+    is configured with (determined by the environment variable
+    `GRAFANA_DASHBOARD_URL`).
+
+    ###### request
+    `GET /api/v1/grafana_dashboard_url`
+
+    ###### response
+    `200` - The body is `{"grafanaDashboardUrl":"$URL"}`.
+
+    `500` - Container JFR is not configured with a Grafana dashboard.
+    Or there was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/api/v1/grafana_dashboard_url
+    {"grafanaDashboardUrl":"http://localhost:3000/d/CazTnbSMz/"}
+    ```
+
+
+* #### `GrafanaDatasourceUrlGetHandler`
+
+    ###### synopsis
+    Returns the URL of the Grafana datasource that Container JFR
+    is configured with (determined by the environment variable
+    `GRAFANA_DATASOURCE_URL`).
+
+    ###### request
+    `GET /api/v1/grafana_datasource_url`
+
+    ###### response
+    `200` - The body is `{"grafanaDatasourceUrl":"$URL"}`
+
+    `500` - Container JFR is not configured with a Grafana datasource.
+    Or there was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/api/v1/grafana_datasource_url
+    {"grafanaDatasourceUrl":"http://localhost:8080"}
+    ```
+
+
+* #### `HealthGetHandler`
+
+    ###### synopsis
+    Returns whether or not the Grafana datasource and Grafana dashbaord
+    that Container JFR is configured with are running properly.
```

Comment:
It also implicitly tells the client whether ContainerJFR itself is running properly, if a response is received vs if it times out or otherwise fails.

Reply:
Mm so what you mean by this is that if the handler returns a `500` or if it times out (would timing out also give a 500?), then something is probably wrong with cjfr because it would mean that cjfr isn't sending/receiving HTTP messages properly?

Reply:
No, if the requesting client gets no response by the end of the timeout period, there would be no response status at all, so no `500` code to interpret. My point here is that if the client gets no response or gets some kind of HTTP(S) request failure, they can also discover information about the status or reachability of ContainerJFR - maybe it is not actually running, or the container is up but somehow the Java process has died or is not responding, or there is a network misconfiguration, etc.

Reply:
Of course, the same conclusions can also be reached if the client sees such failure modes on any path, not just `/health`, but `/health` in particular is a good candidate for this kind of "liveness probe" (https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#when-should-you-use-a-liveness-probe) for clients to check if the container is running properly.